src/ssl.ml

```diff
@@ -301,26 +306,137 @@ end
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Write_error err);
+      end;
+    ret
+
+  external raw_write_substring : socket -> string -> int -> int -> int
+    = "ocaml_ssl_write_blocking" [@@noalloc]
+
+  let write_substring socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.write_substring: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.write_substring: length negative";
+    if start + length > String.length buffer then
+      invalid_arg "Ssl.write_substring: Buffer too short";
+    let ret = raw_write_substring socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Write_error err);
+      end;
+    ret
+
+  external raw_write_bigarray : socket -> bigarray -> int -> int -> int
+    = "ocaml_ssl_write_bigarray_blocking" [@@noalloc]
+
+  let write_bigarray socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.write_bigarray: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.write_bigarray: length negative";
+    if start + length > Bigarray.Array1.dim buffer then
+      invalid_arg "Ssl.write_bigarray: Buffer too short";
+    let ret = raw_write_bigarray socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Write_error err);
+      end;
+    ret
+
+  external raw_read : socket -> Bytes.t -> int -> int -> int
+    = "ocaml_ssl_read_blocking" [@@noalloc]
+
+  let read socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.read: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.read: length negative";
+    if start + length > Bytes.length buffer then
+      invalid_arg "Buffer too short";
+    let ret = raw_read socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Read_error err);
+      end;
+    ret
+
+  external raw_read_into_bigarray :
+    socket -> bigarray -> int -> int -> int =
+    "ocaml_ssl_read_into_bigarray_blocking" [@@noalloc]
+
+  let read_into_bigarray socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.read: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.read: length negative";
+    if start + length > Bigarray.Array1.dim buffer then
+      invalid_arg "Buffer too short";
+    let ret = raw_read_into_bigarray socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Read_error err);
+      end;
+    ret
+
+  external raw_flush : socket -> int =
+    "ocaml_ssl_flush_blocking" [@@noalloc]
+
+  let flush socket =
+    let ret = raw_flush socket in
+    if ret <> 1 then raise (Flush_error(ret = -2))
```

Comment:
this needs a small comment explaining `-2` if a reader hasn't gone through the C stubs yet.

Reply:
done with [f74cf4a](https://github.com/savonet/ocaml-ssl/pull/117/commits/f74cf4a3413addeb6b01da58571869e3d0b320ec)